### DISCUSSION TITLE
`erdtree` package not showing up

### DISCUSSION THREAD

**Discussion by llimllib** [Homebrew/discussions#4381] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.11-52-g97d7a5c
ORIGIN: https://github.com/Homebrew/brew
HEAD: 97d7a5c9b4d55f93fa87ccf33526bbd231ba168a
Last commit: 8 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 6150d43818deed257e0121e1fa6bbbeb1a7e5c14
Core tap last commit: 10 weeks ago
Core tap branch: master
Core tap JSON: 05 Apr 01:30 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: /opt/homebrew/bin/nvim
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.39.0 => /opt/homebrew/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3-arm64
CLT: 14.3.0.0.1.1679647830
Xcode: 14.3
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  pycparser
  trurl
  szip

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  tcl-tk
  python@3.11

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /opt/homebrew/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/opt/homebrew/sbin:$PATH"' >> /Users/llimllib/.bash_profile
```

### Description of issue

On current master, I see that there is an [erdtree formula](https://github.com/Homebrew/homebrew-core/blob/4d34974929e4adca72eea670356c609d75d61a5c/Formula/erdtree.rb)

However, after `brew update`, I can't access the formula.

```
$ brew search erdtree
Error: No formulae or casks found for "erdtree".
```

I see that "core tap last commit" from `brew config` says it was 10 weeks ago - shouldn't `brew update` be updating that? Or is this related to the JSON manifest switch?

Why isn't that package visible to me? What can I do to get it?

**Comment by llimllib** [comment]
I can see that it's present in `https://formulae.brew.sh/api/formula.json`, with no flags that indicate I shouldn't be able to use ti, but I can't currently find out where homebrew stores the tap json it's using.

**Reply by llimllib** [reply]
OK, I found it, it's in `$HOME/Library/Caches/Homebrew/api/formula.jws.json`. Would be cool if that directory was mentioned in `brew config`!

**Reply by llimllib** [reply]
That file contains `erdtree` within the `payload`

**Reply by llimllib** [reply]
the `search` method [here](https://github.com/Homebrew/brew/blob/97d7a5c9b4d55f93fa87ccf33526bbd231ba168a/Library/Homebrew/search.rb#L100) searches all formula names given by `Formula.full_names` [here](https://github.com/Homebrew/brew/blob/97d7a5c9b4d55f93fa87ccf33526bbd231ba168a/Library/Homebrew/formula.rb#L1832), which in turn calls `CoreTap.instance.formula_names` [here](https://github.com/Homebrew/brew/blob/97d7a5c9b4d55f93fa87ccf33526bbd231ba168a/Library/Homebrew/tap.rb#L590), which in turn calls `formula_files.map(&method(:formula_file_to_name))`, which plainly reads the formula files and returns their names.

Since `erdtree` is in the JSON manifest, but doesn't have a formula file (it was added recently, and now `brew update` is updating the manifest but not the formula files), `brew search` doesn't find it

**Reply by llimllib** [reply]
```
$ ls /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/erdtree.rb
ls: /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/erdtree.rb: No such file or directory
```

This is presumably also why `brew doctor` is telling me that `trurl` doesn't have a formula file - it installed it from the JSON manifest but it isn't updating the formula files any longer

**Reply by llimllib** [reply]
Possibly things would start working if I did `brew untap homebrew/core`? But it seems to me that this behavior of not returning any formulas in the API but not in the tap is super broken?

**Reply by llimllib** [reply]
Filed as https://github.com/Homebrew/brew/issues/15153

**Reply by llimllib** [reply]
Search got fixed in https://github.com/Homebrew/brew/pull/15158, thanks for the quick response!

**Comment by llimllib** [comment]
Furthermore, why is `brew doctor` complaining that `trurl` has no formula? I just installed it:

```
$ brew info trurl
==> trurl: stable 0.2 (bottled), HEAD
Command-line tool for URL parsing and manipulation
https://github.com/curl/trurl
/opt/homebrew/Cellar/trurl/0.2 (6 files, 44.8KB) *
  Poured from bottle using the formulae.brew.sh API on 2023-04-04 at 21:31:53
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/trurl.rb
License: curl
==> Dependencies
Required: curl ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 0 (30 days), 0 (90 days), 0 (365 days)
install-on-request: 0 (30 days), 0 (90 days), 0 (365 days)
build-error: 0 (30 days)
```

**Reply by carlocab** [reply]
Do
```
brew untap homebrew/core
```

**Reply by llimllib** [reply]
filed as https://github.com/Homebrew/brew/issues/15159

I am going to do that, but this is still a bug: homebrew is incorrectly reporting these packages as having no formulae because they were installed via the API.

If _having_ the tap present is going to be an issue for people, homebrew should be warning them to untap when they upgrade to a version that supports the API. Otherwise, homebrew should work properly in the presence of a tap that it is no longer updating.

Think of how many people will experience this bug but not report it after they upgrade homebrew

**Reply by carlocab** [reply]
Yes of course it's a bug. Notice that I never said it wasn't.

**Reply by llimllib** [reply]
no worries! apologies if I sounded defensive, unintentional on my part
